--- a/upnphttp.c
+++ b/upnphttp.c
@@ -1398,6 +1398,40 @@
 	return fd;
 }

+#if defined(__APPLE__)
+#define _ICONS_DIR "/Library/Application Support/minidlna/"
+#else
+#define _ICONS_DIR "/usr/share/minidlna/"
+#endif
+
+static char *_load_icon( const char *path, unsigned int *size, unsigned char *alt, unsigned int altsize )
+{
+	FILE *fd;
+	char *ret = NULL;
+	if( path && size )
+	{
+		if( ( fd = fopen( path, "r" ) ) != NULL )
+		{
+			fseek( fd, 0, SEEK_END );
+			*size = ftell( fd );
+			fseek( fd, 0, SEEK_SET );
+			if( *size > 0 )
+			{
+				ret = (char*)malloc( *size );
+				if( ret )
+					fread( ret, *size, 1, fd );
+			}
+			fclose( fd );
+		}
+	}
+	if( !ret )
+	{
+		ret = (char*)alt;
+		*size = altsize;
+	}
+	return ret;
+}
+
 static void
 SendResp_icon(struct upnphttp * h, char * icon)
 {
@@ -1407,32 +1437,61 @@
 	int size;
 	struct string_s str;

+	static char *_icons[4]= {0};
+	static unsigned int _sizes[4]= {0};
+	if( *_sizes == 0 ) {
+		_icons[0] = _load_icon( _ICONS_DIR "small.png", _sizes + 0, png_sm, sizeof( png_sm ) - 1 );
+		_icons[1] = _load_icon( _ICONS_DIR "large.png", _sizes + 1, png_lrg, sizeof( png_lrg ) - 1 );
+		_icons[2] = _load_icon( _ICONS_DIR "small.jpg", _sizes + 2, jpeg_sm, sizeof( jpeg_sm ) - 1 );
+		_icons[3] = _load_icon( _ICONS_DIR "large.jpg", _sizes + 3, jpeg_lrg, sizeof( jpeg_lrg ) - 1 );
+	}
+
 	if( strcmp(icon, "sm.png") == 0 )
 	{
 		DPRINTF(E_DEBUG, L_HTTP, "Sending small PNG icon\n");
-		data = (char *)png_sm;
-		size = sizeof(png_sm)-1;
+		if( _sizes[0] ) {
+			data = _icons[0];
+			size = _sizes[0];
+		} else {
+			data = (char *)png_sm;
+			size = sizeof(png_sm)-1;
+		}
 		strcpy(mime+6, "png");
 	}
 	else if( strcmp(icon, "lrg.png") == 0 )
 	{
 		DPRINTF(E_DEBUG, L_HTTP, "Sending large PNG icon\n");
-		data = (char *)png_lrg;
-		size = sizeof(png_lrg)-1;
+		if( _sizes[1] ) {
+			data = _icons[1];
+			size = _sizes[1];
+		} else {
+			data = (char *)png_lrg;
+			size = sizeof(png_lrg)-1;
+		}
 		strcpy(mime+6, "png");
 	}
 	else if( strcmp(icon, "sm.jpg") == 0 )
 	{
 		DPRINTF(E_DEBUG, L_HTTP, "Sending small JPEG icon\n");
-		data = (char *)jpeg_sm;
-		size = sizeof(jpeg_sm)-1;
+		if( _sizes[2] ) {
+			data = _icons[2];
+			size = _sizes[2];
+		} else {
+			data = (char *)jpeg_sm;
+			size = sizeof(jpeg_sm)-1;
+		}
 		strcpy(mime+6, "jpeg");
 	}
 	else if( strcmp(icon, "lrg.jpg") == 0 )
 	{
 		DPRINTF(E_DEBUG, L_HTTP, "Sending large JPEG icon\n");
-		data = (char *)jpeg_lrg;
-		size = sizeof(jpeg_lrg)-1;
+		if( _sizes[3] ) {
+			data = _icons[3];
+			size = _sizes[3];
+		} else {
+			data = (char *)jpeg_lrg;
+			size = sizeof(jpeg_lrg)-1;
+		}
 		strcpy(mime+6, "jpeg");
 	}
 	else
